core/src/main/scala/kafka/raft/KafkaMetadataLog.scala

```diff
@@ -147,18 +221,102 @@ class KafkaMetadataLog(
   }
 
   override def createSnapshot(snapshotId: raft.OffsetAndEpoch): RawSnapshotWriter = {
-    FileRawSnapshotWriter.create(log.dir.toPath, snapshotId)
+    // Do not let the state machine create snapshots older than the latest snapshot
+    latestSnapshotId().ifPresent { latest =>
+      if (latest.epoch > snapshotId.epoch || latest.offset > snapshotId.offset) {
+        // Since snapshots are less than the high-watermark absolute offset comparison is okay.
```

Comment:
Is it useful here to ensure that `snapshotId` here is lower than the high watermark?